### ISSUE TITLE
Is there a way to store/view predictions | In order to analyze example results

### ISSUE THREAD

**Issue by sigadavid96** [zjunlp/DART#6]
Hi,

Great work!
I'm currently looking at the paper and running experiments on it. 
In the codes: **Is there a way to store/view predictions, In order to analyze example results?**

I wanna see what are the shortcomings of regular techniques and how DART has overcome them, 
and lastly, are there still things left for DART to solve or opportunities to improve DART?

Thank you

**Comment by Riroaki** [comment]
Hi, thanks for your attention!
The predictions can be produced & stored using `inference.py`; FYI, `visualize.py` / `visualize_word_emb.py` contains some visualization codes of word embedding / last hidden states output.
Currently we've been working on a clean branch for code refraction and apply some more training techniques to improve model performance.

**Comment by dsiga96** [comment]
Thank you!
I was wondering if there's an existing implementation code for BERT instead of RoBERTa.
I was also wondering about what are the challenges we might face if we try other pre-trained models. 
Do you think DART could be compatible to all the BERT bases models?

Best Regards,
Davidson

**Comment by dsiga96** [comment]
Also, is there a good way to reach you via E-mail?
Thank you!

**Comment by Riroaki** [comment]
> Thank you! I was wondering if there's an existing implementation code for BERT instead of RoBERTa. I was also wondering about what are the challenges we might face if we try other pre-trained models. Do you think DART could be compatible to all the BERT bases models?
> 
> Best Regards, Davidson

Hi Davidson, thanks for you attention! In current codes we use `AutoModelForMaskedLM` which supports multiple BERT-like LMs. If you have any problems using other LMs you may open an issue and we'll see it.

**Comment by sigadavid96** [comment]
Hi!
Thank you for all the comments.
I had another query.
In codes, I'm trying to find the lambda hyperparameter which was defined in the paper. The Loss1+Lambda*Loss2
I was trying to find where this lambda hyperparameter was coded, and how we could experiment with it.

Additionally, 
The selection of the initial prompts like "It was" for SST-2 was hardcoded (of course, to initialize), but I was wondering how you came up with the possibilities of prompts available for Initialization, is it purely from prior papers?

Thank you!

**Comment by Riroaki** [comment]
> 

Hi sigadavid96, for your first question, you may check the current version of codes and see the hyper-parameters in `config/sample.yml`. The lambda refers to `mlm_loss_weight` in our codes:
https://github.com/zjunlp/DART/blob/0ed51c05be7e0b88e6ce91c224b2889adc61bc28/config/sample.yml#L19

For your second question, we use former papers prompts for initialization and also tried out randomized searching. It turns out the former ones are better in practice, and you can modify the prompts in the data Reader classes in `src/data.py`.

**Comment by sigadavid96** [comment]
Oh! Looks like the codes have been refactored! These look great! Thank you!
Also, thank you for the links and the answers to the questions.
It looks like 
mlm_loss_weight: 0.0  # weight of auxiliary MLM loss 
Lambda has been initialized to 0, is the benchmark DART using this as the lambda too?

Thank you!

**Comment by Riroaki** [comment]
Hi, in our experiment we followed LM-BFF and searched optimal hyper parameters on each data split for every GLUE dataset respectively; for most cases, setting mlm loss weight to 0 helps reaching better results, and it depends.
You may search on wandb.ai with the scripts `sweep.py` for your own experiments.

**Comment by lancorrect** [comment]
Hi, 
according to your reply, mlm loss weight was set to 0. Does it mean that Fluency Constraint Object mentioned in paper was not used through the entire inplemententation?
And I caould not see `visualize.py` or `visualize_word_emb.py` in current repo. Could you please tell me how I can visualize the representation of the masked tokens?
Thanks a LOT!

**Comment by Riroaki** [comment]
Hi, in our experiment the auxiliary loss is used and the results won't differ much on GLUE with/without the fluency object; in our project the weight is set to 0.0 in our `sample` configuration just for debugging convenience.

In the recent update the visualization codes are removed for simplicity; anyway, I will add them back soon and notify you in this issue when available!

**Comment by lancorrect** [comment]
Thanks for your fast reply! I appreciate it.

If the weight is set to 0.0 just for debugging convenience, could I know what is the optimal mlm experiment weight? Because It was not mentioned in your paper's appendix.

And I'm looking forward to your visualization codes. Thanks again for notifying me at that time!

**Comment by Riroaki** [comment]
As we mentioned in the paper, the performance is evaluated on each data split and therefore the optimal loss weight varies as other hyper parameters; I suggest 0.05 worth a try and you can perform fine-grained search on different datasets with `sweep.py`.
